archiver/archive.go

```diff
@@ -0,0 +1,143 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package archiver
+
+import (
+	"archive/tar"
+	"compress/gzip"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+
+	"github.com/joeshaw/multierror"
+
+	"github.com/pkg/errors"
+)
+
+// PackageProperties defines properties describing the package. The structure is used for archiving.
+type PackageProperties struct {
+	Name    string
+	Version string
+	Path    string
+}
+
+// ArchivePackage method builds and streams an archive with package content.
+func ArchivePackage(w io.Writer, properties PackageProperties) (err error) {
+	gzipWriter := gzip.NewWriter(w)
+	tarWriter := tar.NewWriter(gzipWriter)
+	defer func() {
+		var me multierror.Errors
+
+		if err != nil {
+			me = append(me, err)
+		}
+
+		err = tarWriter.Close()
+		if err != nil {
+			me = append(me, errors.Wrapf(err, "closing tar writer failed"))
+		}
+
+		err = gzipWriter.Close()
+		if err != nil {
+			me = append(me, errors.Wrapf(err, "closing gzip writer failed"))
+		}
+
+		if me != nil {
+			err = me.Err()
+		}
+	}()
+
+	rootDir := fmt.Sprintf("%s-%s", properties.Name, properties.Version)
+
+	err = filepath.Walk(properties.Path, func(path string, info os.FileInfo, err error) error {
```

Comment:
Out of curiosity: Will this follow symlinks?

Reply:
I will quote the docs here:

>// Walk does not follow symbolic links.